File: packages/cli/src/commands/parse.ts
import type { Command } from 'commander';
import { parseVaa } from '@worm-tool/sdk';
import { printJson, printError } from '../output.js';

export function registerParseCommand(program: Command): void {
  program
    .command('parse <vaa>')
    .description('Parse a VAA from hex or base64 and print its fields as JSON')
    .action((vaa: string) => {
      try {
        printJson(parseVaa(vaa));
      } catch (err) {
        printError('Failed to parse VAA', err);
        process.exit(1);
      }
    });
}

File: archive/rust-crates/wormhole-cli/src/commands/parse.rs
use clap::Args;
use wormhole_sdk::vaa;

use crate::output;

/// Parse a VAA and print its fields as JSON.
#[derive(Debug, Args)]
pub struct ParseArgs {
    /// Hex (with or without 0x prefix) or base64-encoded VAA.
    pub vaa: String,
}

/// Run the `worm parse` command.
///
/// # Errors
///
/// Returns an error if the VAA cannot be decoded or printed.
pub fn run(args: &ParseArgs) -> anyhow::Result<()> {
    let vaa_data =
        vaa::parse(&args.vaa).map_err(|e| anyhow::anyhow!("failed to parse VAA: {e}"))?;
    output::print_json(&vaa_data)
}
